### ISSUE TITLE
Add Mastodon support

### ISSUE THREAD

**Issue by AstraLuma** [pennersr/django-allauth#3218]
https://docs.joinmastodon.org/spec/oauth/

**Comment by AstraLuma** [comment]
I might see about doing a version of this myself, if nobody gets to it first.

**Comment by pennersr** [comment]
Not sure if Mastodon is OIDC compliant. If so, you can already use the OIDC functionality introduced in the latest version. If not, feel free to submit a pull request.

**Comment by AstraLuma** [comment]
No, mastodon does not support OpenID Connect. I linked to the reference document above.

The process is to automatically register an OAuth app and then use OAuth2.

**Comment by AstraLuma** [comment]
Ok, correction:

I'm going to have to bastardize OpenID Connect, since the UX and backend flows aren't entirely dissimilar. If you squint. And take off your glasses.

**Comment by adamghill** [comment]
I took a look at this for a little bit, but couldn't figure out how to force the Mastodon flow into `django-allauth` especially because the available servers is dynamic. So, I ended up writing the oauth flow myself. It needs some more testing and handling of error conditions, but the basic flow seems to work.

The code is here: https://github.com/adamghill/fediview/tree/main/account. I was thinking of breaking it out into it's own library, though.
